### DISCUSSION TITLE
Schema awareness: can I avoid using it? How can I optimize schema.json? Can I code-split it?

### DISCUSSION THREAD

**Discussion by frederikhors** [urql-graphql/urql#974] [Q&A]
# Summary

- [ ] Use URQL CacheExchange without Schema awareness

  - [ ] Fix the infinite reloading issue

  - [ ] Document how to

- [ ] Document ways we can optimize Introspection Schema

  - [x] Amazing tool from urql team (https://github.com/FormidableLabs/urql/pull/976)

    - [ ] Strip out "descriptions" from final `.json`

    - [ ] Sort alphabetically (or such) final `.json` to check for any changes

  - [ ] Document how to wrap the entire introspection result in a string and `JSON.parse(it)`; update example (maybe the svelte one) with this
  
  - [ ] Is there something else to remove from the final schema?

- [ ] Generate schema

  - [x] Generate it from API introspection call (docs OK!)

  - [ ] Generate it from `*.graphql` files (needs script)

- [ ] How to load schema the async way with code-splitting, docs

## Story

Schema awareness is the most loved and hated thing for those who use URQL.

<details>
<summary>Read the long story</summary>

As soon as the application becomes a little bigger than a simple "hello world" you need to use different queries for the same entities that you can first see as part of a list and then as a detail.

Some fields of these queries may be `null` but URQL cannot know this. He needs someone to "show him the way".

That someone is our GraphQL Schema.

One advantage of the awareness schema is that urql doesn't have to wait for new query data to fetch and can show the data it has already cached for that entity.

One "disadvantage" is that for some projects the schema is very large and this is a problem both for the final bundle size and for the browser's parsing time (TTI, time to interactive).

</details>

## Use URQL CacheExchange without Schema awareness:

- [We started talking about this here](https://spectrum.chat/urql/help/how-about-a-convention-for-list-detail-that-speeds-up-the-ui-a-lot~4bcb6019-8ada-4b6d-b325-72e9244217d8)

- @kitten suggested a method that might work, but he also made it clear that it is not "sponsored" by the URQL team because quite rightly it is error-prone and may not make 100% sense:

  <details>
  <summary>See the suggested code</summary>

  ```js
  const maybePartial = (parent, args, cache, info) => {
    const fieldValue = cache.resolve(info.parentKey, info.fieldName, Object.keys(args).length ? args : undefined);
    if (fieldValue === null) {
      info.partial = true;
      return null;
    } else {
      return fieldValue;
    }
  };

  const resolvers = {
    Query: {
      todo: (parent, args, cache, info) => ({ __typename: "Todo", id: args.id })
    },
    Todo: {
      details: maybePartial
    }
  };
  ```
  </details>

  You can see a [demo here](https://codesandbox.io/s/urql-svelte-crud-list-detail-different-queries-forked-lqukh)

- the problem with this is that when backend server returns `null` values for some values Urql keeps fetching API

  <details>
  <summary>See the reproduction in this GIF</summary>

    ![c70ef953-6c58-4be4-9032-54c9130b843d-issue](https://user-images.githubusercontent.com/41120635/92924246-ed038c00-f438-11ea-83cf-b99de29eea4c.gif)
  </details>

  - **we are waiting for an answer to this**

## Document ways we can optimize Introspection Schema (if we must use it):

- [We talked about it here](https://spectrum.chat/urql/help/any-advice-on-reducing-the-size-of-the-schema-file-that-needs-to-be-bundled-with-the-app~75c85a8d-3919-436a-bcdf-54992aa6f906)

- The ideas suggested and that - if shared by the team - should be documented are:

  - @earksiinni proposed to remove fields like: `defaultValue`, `isDeprecated`, `deprecationReason`; I think it is wrong to remove them tout court; at most we could assume a default value in URQL for these fields and only then remove from the json those that are not our default. What do you think?

  - @kitten suggested to remove `directives` key from json; it works

  - @kitten needs to test if we can remove every type that starts with `__` in `schema.__types`

  - @kitten:
      > There's also (if you're using an Apollo Server) some CacheControlScope and related types in your schema export, which are also superfluous

      > In case the internal GraphQLSchema object relies on some __* hints, which would actually be rather weird, we can likely add some code to patch it up

      > if you're concerned about the parsing time (good hint on that) I'd measure it, and then wrap the entire introspection result in a string and JSON.parse(...) which is typically cheaper and faster for large JSON blobs

  ### How to generate schema: from server or from files?

  - We can use the [instructions here](https://formidable.com/open-source/urql/docs/graphcache/schema-awareness/) for server introspection query

  - or we can use [this plugin for graphql-code-generator](https://graphql-code-generator.com/docs/plugins/introspection) which use our `*.graphql` files, which is sometimes the only way you can use if you don't have the server started (and with introspection active) (maybe in CI/CD envs); today with the server way my schema.json is 30% lighter; I still have to understand why

## How to load it the async way with code-splitting

- I'm using Svelte, so my problem is with Svelte, but I think this is good for everyone to have it in the docs

- I'm using the [svelte example in your packages folder here](https://github.com/FormidableLabs/urql/blob/main/packages/svelte-urql/example/src/App.svelte) except I'm creating a `format: 'esm'` in my `rollup.config.js`

  - my Rollup is already code-splitting my final bundle in chunks when I use `import('./Component.svelte')` in my code

  - What I'm having hard time doing is initializing urql asynchronously, like this:

    ```js
    import { initClient, dedupExchange, fetchExchange } from '@urql/svelte'
    import { cacheExchange } from "@urql/exchange-graphcache";

    const exchanges = [
      cacheExchange({
        resolvers: { /*... my resolvers*/ },
        schema // I need this async, something like: await import('./schema').then(result => result.default)
      }),
      fetchExchange
    ]

    export function initURQL () {
      initClient({
        url: 'http://localhost/graphql',
        exchanges
      })
    }
    ```

  - if I use something like

    ```js
    export async function initURQL () {
      const schema = await import('./schema.json').default

      const exchanges = [
      cacheExchange({
        resolvers: {},
        schema
      }),
      fetchExchange
      ]
      initClient({
        url: 'http://localhost/graphql',
        exchanges
      })
    }`
    ```

    and call it from `App.svelte` with `;(async () => await initURQL())()` it gives me error:

    ```console
    Uncaught (in promise) Error: Function called outside component initialization
    ```

  - **does it make sense to introduce a new async method in URQL to get schema in async way?**

---

**Thanks for your great work and your time!**

**Comment by kitten** [comment]
I personally don't think it makes sense to async load the schema as it'd block all requests eliminating any TTI benefits. The reality is that your application with SSR will probably have a good TTFB and 10KB JSON is not going to impact this significantly.

We assume that schema awareness really pays off for larger apps with lots of interactivity, so skipping that shouldn't have a major impact apart from several loads having to happen before data is displayed, which is expected for low-interactive pages.

I think before making any assumptions we are planning to test the size of JSON schemas and how they zip. Parsing times are pretty irrelevant because JSON is extremely cheap to be parsed. Do we're mainly focusing on size here.

We can always consider the manual escape hatch as an alternative, but until then it'll go undocumented 😅

**Comment by frederikhors** [comment]
I updated the first post with a summary (from my perspective).

Here some ideas for introspection's schema utilities:

1. I think we should have a utility to **generate the schema from the `*.graphql` files** and not just by calling the server (CI/CD environments).

    I know we can already use `codegen`, but there are issues with it:

    - the "descriptions" are generated (I also opened this issue about it: https://github.com/dotansimha/graphql-code-generator/issues/4731 and I await)
    - we need to use an external tool (more dependencies, more chaos)

1. We can also handle (**strip out**) **"descriptions"** in our `minifyIntrospectionQuery` code (also useful for when you are other ways to retriev it - instead of codegen - and you can't prematurely remove descriptions from the .json file).

1. Everytime I use [the script in docs](https://formidable.com/open-source/urql/docs/graphcache/schema-awareness/) the server gives me an answer (and therefore a file) always different from the previous call. I noticed that what changes is only the order of the information inside.

    **Can we sort alphabetically** (or by another criterion) the final json in order to check for any differences automatically (CI/CD environments)?

**Comment by kitten** [comment]
1. That's mainly other tools' responsibilities, but this is really easily accomplished (see the second helper in that package)
2. ~Agreed, PRs welcome, but for the current instructions it does work~
3. Also not part of what we do 😅 But also possible to be added to the utility 

Edit: `descriptions` will already be stripped out. I just forgot that that was the case

**Reply by bbigras** [reply]
>     1. That's mainly other tools' responsibilities, but this is really easily accomplished (see the second helper in that package)

Which helper is that?

**Reply by kitten** [reply]
@bbigras If you're not already using it, GraphQL Code Generator makes it easier to generate schema JSON files for your app.

**Reply by bbigras** [reply]
Yes I'm using it, but my schema is huge. I was wondering if it was possible to generate my schema with only the types that my queries are using (to have a smaller schema). I think I misinterpreted what "I think we should have a utility to generate the schema from the *.graphql files and not just by calling the server (CI/CD environments)." meant.

**Reply by PierreAndreis** [reply]
@bbigras i think this responsibility would fall into codegen plugin https://www.graphql-code-generator.com/docs/plugins/urql-introspection and not into urql library itself

**Reply by kitten** [reply]
@bbigras it definitely is possible to filter down your schema as long as there are parts you're never using; but that'd be currently something you'd have to implement. But the introspection schema data is really simple to work with. So if you take a browse at `@urql/introspection`'s code you can see that filtering by type for instance is pretty simple
